priv/samples/basics/spi_rgb_leds.livemd

```diff
@@ -0,0 +1,371 @@
+    end
+    Process.sleep(delay)
+end)
+```
+
+[return to top](#table-of-contents)
+
+## Managing all the LEDs on a strip
+
+We can create lists of frames and send them all at one.
+
+Update the `led_count` to reflect your strip.
+
+```elixir
+# Set the number of LED's in your strip
+led_count = 144
+
+# The end frame is 8 '0' bits * (leds - 1) / 16 (rounded up)
+# These are clock edges to ensure the LED's respond immediately and is borrowed from the FASTLed library.
+end_frame =
+  Enum.map(1..round((led_count - 1) / 16), fn _ -> <<0::8>> end)
+  |> Enum.map_join(& &1)
+
+# Creating a List of frames with each LED on white
+led_frames = Enum.map(1..led_count, fn _ -> led_frame end)
+off_frames = Enum.map(1..led_count, fn _ -> off_frame end)
+
+# Transfer the frames over spi
+Circuits.SPI.transfer(ref, start_frame <> Enum.join(led_frames) <> end_frame)
+
+# Wait for a prespecified duration of time.
+Process.sleep(delay)
+
+# Turn them off
+Circuits.SPI.transfer(ref, start_frame <> Enum.join(off_frames) <> end_frame)
+```
+
+[return to top](#table-of-contents)
+
+## Simple larson scanner
+
+A larson scanner, named after Glen A. Larson, is a back and forth scanning red light.
+
+We are going to create a map the length of our LED strip for addressing.
+
+```elixir
+delay = 1
+scans = 2
+
+# Create an addressible Map of all your strips LED's.
+map = Enum.reduce(0..led_count, %{}, fn x, acc -> Map.put(acc, x, off_frame) end)
+
+# For each scan
+Enum.map(1..scans, fn _ ->
+  # Write from left to right
+  Enum.map(0..(led_count - 1), fn x ->
+    map = Map.replace(map, x, <<7::3, 15::5, 0x00, 0x00, 0x0f>>)
+    bytes = 
+      map 
+      |> Enum.sort 
+      |> Enum.map(fn {_, v} -> v end) 
+      |> Enum.join
+    Circuits.SPI.transfer(ref, start_frame <> bytes <> end_frame)
+    Process.sleep(delay)
+  end)
+
+  # Write from right to left
+  Enum.map((led_count - 1)..0, fn x ->
+    map = Map.replace(map, x, <<7::3, 15::5, 0x00, 0x00, 0x0f>>)
+    bytes = 
+      map 
+      |> Enum.sort 
+      |> Enum.map(fn {_, v} -> v end) 
+      |> Enum.join
+    Circuits.SPI.transfer(ref, start_frame <> bytes <> end_frame)
+    Process.sleep(delay)
+  end)
+end)
+
+# Turn them off
+Circuits.SPI.transfer(ref, start_frame <> Enum.join(off_frames) <> end_frame)
+```
+
+[return to top](#table-of-contents)
+
+## Simple rainbow
+
+```elixir
+scans = 1
+delay = 1
+
+# A quick and dirty guard for lower and upper limits.
+min_max = fn
+  (_min, max, value) when value > max -> max
+  (min, _max, value) when value < min -> min
+  (_min, _max, value) -> value
+end
+
+# A conversion function from HSV to RGB.
+from_hsv = fn(h, s, v) ->
+    h = min_max.(0, 360, h)
+    s = min_max.(0, 100, s)
+    v = min_max.(0, 100, v)
+
+    h = h / 60
+    i = Float.floor(h) |> trunc()
+    f = h - i
+    sat_dec = s / 100
+    p = v * (1 - sat_dec)
+    q = v * (1 - sat_dec * f)
+    t = v * (1 - sat_dec * (1 - f))
+    p_rgb = (p * 255 / 100) |> trunc()
```

Comment:
```suggestion
    p_rgb = floor(p * 255 / 100)
```
My C roots prefer calling `floor`. If you like `trunc/1` better, then that's fine too.

Reply:
`floor` looks much nicer. Updated.